### ISSUE TITLE
Add QueryParamDecoder helpers to work with fallible decoders

### ISSUE THREAD

**Issue by kubukoz** [http4s/http4s#1997]
I just did

```scala
import java.time.ZoneId

val decoder: QueryParamDecoder[ZoneId] = value => QueryParamDecoder[String].decode(value)
  .andThen { str =>
    Validated.catchNonFatal(ZoneId.of(str))
      .leftMap(ex => ParseFailure(str, ex.getMessage))
      .toValidatedNel
  }
```

to avoid a server-side error on malformed input. As far as I know, there's no better way of doing this. I think that might be a common enough example (interacting with APIs that throw, or return `ValidatedNel[ParseFailure, A]`) - and would like to suggest adding some helper, like

```scala
//in trait QueryParamDecoder[T]
def emap[U](f: T => ValidatedNel[ParseFailure, U]): QueryParamDecoder[U]
```

This would make the above a bit more concise:

```scala
val decoder: QueryParamDecoder[ZoneId] = QueryParamDecoder[String].emap(str => Validated.catchNonFatal(...).leftMap(...).toValidatedNel)
```

but it would still require the hassle of `catchNonFatal` + mapping the exception to a `ParseFailure` + `toNel`. We could also add a type alias for `ValidatedNel[ParseFailure, T]` (or a newtype), and provide `catchNonFatal` on its companion object. The above snippet would now be:

```scala
val decoder: QueryParamDecoder[ZoneId] = QueryParamDecoder[String].emap(str => QueryParamParseResult.catchNonFatal(ZoneId.of(str)))
```

I'd love to hear some feedback about this :)

**Comment by rossabaker** [comment]
I'm :+1: on a function like `emap`.  Is [circe](https://github.com/circe/circe/blob/a5973d360d4c04b70ee4d14e435430e2aa32e85d/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L187) your inspiration here?  Any other prior art?  I'd like to get the name right.

Type aliases are something we have historically come to regret.  I'm a slight :-1: on that.

I don't think we're changing any instances, so I'm :-1: on a newtype.  Circe's [`emapTry`](https://github.com/circe/circe/blob/a5973d360d4c04b70ee4d14e435430e2aa32e85d/modules/core/shared/src/main/scala/io/circe/Decoder.scala#L216) looks close to what you're trying to accomplish with the `catchNonFatal`.

**Comment by kubukoz** [comment]
Yeah, circe was part of my inspiration. It also has a type alias for the decoding result.

`emapTry` looks a bit like what I need, but auto-catching would be even more handy... although I don't need it that often, after all.

BTW in this particular usecase I dropped `QueryParamDecoder` and went for directly extracting the parameter from the request object.

**Comment by ybasket** [comment]
I just stumbled upon the same problem and would in high favour of adding the non-auto-catching version (can also do a PR). As for the auto-catching, I'm not sure that's a good idea.

Naming: `emap` is for me bound to `Either` (as circe's `Decoder` has it that way and `e` in front indicates) - though `emapTry` (circe) adds the `Try` in the end…maybe `emapValidated` then and `emap` for plain `Either`?

**Comment by rossabaker** [comment]
Fixed by #2656.
